src/rmt.rs

```diff
@@ -0,0 +1,332 @@
+//! implementation. It is recommended to read before using this module.
+//!
+//! This is an initial implementation supporting:
+//!  * Transmission, currently only blocking.
+//!
+//! Not supported:
+//! * Receiving.
+//! * No channel locking. (To prevent users from accidentally reusing a channel.)
+//! * No buffer protection while transmitting.
+//! * Change of config after initialisation.
+//!
+//! # Example Usage
+//! ```
+//! use esp_idf_hal::gpio::Output;
+//! use esp_idf_hal::rmt::Channel::Channel0;
+//!
+//! let peripherals = Peripherals::take().unwrap();
+//! let led: Gpio18<Output> = peripherals.pins.gpio18.into_output().unwrap();
+//! let config = WriterConfig::new(&led, Channel0).clock_divider(1);
+//! let mut writer = Writer::new(config).unwrap();
+//!
+//! writer.push_pulse(Pulse::new(PinState::High, PulseTicks::max()));
+//! writer.push_pulse(Pulse::new(PinState::Low, PulseTicks::max()));
+//!
+//! writer.start().unwrap();
+//!
+//!```
+//!
+//! See the `examples/` folder of this repository for more.
+
+// TODO: Do we need to prevent users from creating two drivers on the same channel?
+// TODO: Should probably prevent writing to buffer while rmt is transmitting:
+//       "We must make sure the item data will not be damaged when the driver is still sending items in driver interrupt."
+use crate::gpio::OutputPin;
+use embedded_hal::digital::PinState;
+use esp_idf_sys::{
+    esp, rmt_channel_t_RMT_CHANNEL_0, rmt_channel_t_RMT_CHANNEL_1, rmt_channel_t_RMT_CHANNEL_2,
+    rmt_channel_t_RMT_CHANNEL_3, rmt_config, rmt_config_t, rmt_config_t__bindgen_ty_1,
+    rmt_driver_install, rmt_get_counter_clock, rmt_item32_t, rmt_item32_t__bindgen_ty_1,
+    rmt_item32_t__bindgen_ty_1__bindgen_ty_1, rmt_mode_t_RMT_MODE_RX, rmt_mode_t_RMT_MODE_TX,
+    rmt_tx_config_t, rmt_write_items, EspError, RMT_CHANNEL_FLAGS_AWARE_DFS,
+};
+use std::mem::ManuallyDrop;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[repr(u32)]
+pub enum Channel {
+    // TODO: Work out the different number of channels per chip.
+    Channel0 = rmt_channel_t_RMT_CHANNEL_0,
+    Channel1 = rmt_channel_t_RMT_CHANNEL_1,
+    Channel2 = rmt_channel_t_RMT_CHANNEL_2,
+    Channel3 = rmt_channel_t_RMT_CHANNEL_3,
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[repr(u32)]
+pub enum Mode {
+    Tx = rmt_mode_t_RMT_MODE_TX,
+    Rx = rmt_mode_t_RMT_MODE_RX,
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct Pulse {
+    ticks: PulseTicks,
+    pin_state: PinState,
+}
+
+impl Pulse {
+    pub fn new(pin_state: PinState, ticks: PulseTicks) -> Self {
+        Pulse { pin_state, ticks }
+    }
+}
+
+pub struct WriterConfig {
+    config: rmt_config_t,
+}
+
+impl WriterConfig {
+    pub fn new<OP>(pin: &OP, channel: Channel) -> Self
+    where
+        OP: OutputPin,
+    {
+        // Defaults from https://github.com/espressif/esp-idf/blob/master/components/driver/include/driver/rmt.h#L101
+        Self {
```

Comment:
> turn WriterConfig into a lightweight Rust config object

I did initially do this, but read about embedded-hal suggesting implementations should be zero copy so I refactored. Will change it back.

Reply:
Interesting, can you point me at a link describing that?

Reply:
After reading it again, it doesn't specifically say implementations should follow the projects "Design goals". To me it implies it, but I'm not sure.

https://docs.rs/embedded-hal/latest/embedded_hal/

> Must be minimal, and thus easy to implement and zero cost, yet highly composable. People that want higher level abstraction should prefer to use this HAL rather than re-implement register manipulation code.

Reply:
> perhaps just name it Config then.

I named it `WriterConfig` because this is the tx side of the rmt implementation and it has a fairly different configuration to rx. Maybe we can separate writing and reading into a module? e.g. rmt::writer::config::Config ?

Also, I picked "write" because it seems to be the standard in Rust I/O, but IMO it makes less sense in embedded. Totally happy to change the names around if something sounds better, e.g. tx, send, output, etc. Something like rmt::tx::config::Config seems nicer to me.

Reply:
It talks about "zero cost", not about "zero copy". Regardless, your interpretation of "zero cost" to include the one-time configuration structure is perhaps going a bit too far. I would be much more worried about the other potential performance issues (the FP operations; the Vec that requires `alloc` - see comments below) than this one.

As for `WriterConfig` - sure, let's keep it that way, until the rest is resolved, low prio anyway.

Reply:
> one-time configuration structure is perhaps going a bit too far

You're absolutely right. I have a bad case of premature optimisation.